damian-lis/checkout-flow
# CF-01 Initial setup

# CH-01 Saleor and initial repo setup.

This is the first task regarding the development of the ‘checkout-flow’ app.

Before diving into the app's functionalities, we need to set up the Saleor platform. This platform will provide the essential backend logic needed to create products, channels, shipping methods, and so on.

In addition to that we should also prepare our development environment to facilitate the coding process and set up basic configuration for unit/integration tests that will be used during creating app logic.

**ACs:**

- A new Saleor instance is created
- the ‘checkout-flow’ app is set up with:
    - Next.js version 13
    - TypeScript
    - TailwindCSS
    - Apollo Client which is connected to the Saleor instance for GraphQL operations
        - the codegen tool is used to automate process of generating code based on the GraphQL schema
    - some linting, prettier, and husky rules for code consistency
- the 'checkout-flow' app is deployed on Vercel with an example of client and server components that fetch some basic product data

**Potential problems:**

- I'm unfamiliar with Saleor, so I'll need to delve into its documentation to understand the setup process for the aspects mentioned in the ACs.

**Time estimation:** 180 min

- deploying to Vercel (15 min)
- the Saleor instance setup:
    - read the documentation to become familiar with it (30 min)
    - create one product, one channel, and one shipping method (30 min)
- the 'checkout-flow' app setup:
    - install Next.js version 13 with TypeScript and Tailwind CSS (15 min)
    - set up Apollo Client, connect it to the Saleor instance, and integrate the codegen tool (60 min)
    - configure linting, Prettier, and Husky rules (30 min)

**Result:**
The development time was a little bit longer due to some unexpected behaviors that occurred while setting up the project configuration.

<hr/>

The following is a DERBY document regarding the project (was not changed during and after development):

# DERBY

**Problem**

Create a simple checkout flow that allows a user to buy one product.

**Functional requirements**

- implement an action allowing users to initiate the checkout process with predefined one product.
- create forms designed to collect the essential information required for order placement:
    - personal details: user's name and email
    - shipping address
    - billing address, along with the Stripe payment method
- create a summary section showcasing the breakdown and total calculation for the checkout
- create an order as a result of completing checkout

**Non-functional requirements**

- ensure a positive user experience by displaying loaders or spinners during backend response wait times.
- implement a responsive design, ensuring content is easily readable and the layout structure looks good

**Proposed solution**

To solve the problem in question we can use:

- Next.js v13 to build a server-side rendered application for optimal SEO and performance,
- Saleor to provide the e-commerce backend for a checkout flow,
- Stripe for managing a user payment,
- React hook form with ZOD to streamline form validations,
- Apollo Client to manage GraphQL interactions with the Saleor backend,
- TailwindCSS as a utility-first CSS framework for writing styles in a faster way (without class naming fatigue).

**New components**

- Button for performing actions such as saving form data or starting a checkout process,
- Section for displaying form fields, saved form data, and checkout summary,
- InputField for form details, used in sections like contact, shipping, and billing,
- SelectField for choosing options, like selecting countries,
- AddressFields for managing address data that consist of InputFormField and SelectFormField components. It will be used both in the shipping address and billing address sections,
- StripeComponent as a wrapper for the Stripe SDK for managing payments.

**Failure cases and remediation**

- error while creating a checkout
    - display a user-friendly error message that to inform a user what happened
- error while saving form data on the checkout page
    - display a user-friendly error message below the field where the error is related to
        - if an error is generic and is not related to a particular field, display an error notification above the section action button to inform a user what happened

**Alternative solutions**

As alternative technologies, you can use:

- NextJS v12 over v13 for stability reasons (the 13 version just recently came out),
- the Woocommerce platform over the Saleor one due to the fact it might be more mature and has global recognition,
- the Adyen payment process over the Stripe one
- the yup library instead of ZOD one,
- traditional REST API over GraphQL since it is widely used and many developers are familiar with it,
- SCSS modules the styled-components library since it doesn’t require learning new ways of writing styles.

**Dependencies**

- Saleor for e-commerce backend.
- Stripe for payment processing.
- Next.js a framework for server-side rendering.
- Apollo Client for managing GraphQL interactions.
- TailwindCSS: for styling and responsiveness.
- React Hook Form for form validation

**QA Validation** 

- ensure a user can initiate the checkout process by clicking the button on the initial page,
- check the frontend form field validation in each section,
- check a display of error messages received from the backend,
- check if the data provided in each section is saved correctly,
- ensure the saved checkout data remains the same after a page reload,
- test dynamic shipping cost calculation based on data provided in the shipping section,
- check the process of placing an order,
- check the design's responsiveness on various devices,
- ensure sections can be collapsed when needed.
